### ISSUE TITLE
TrustServerCertificate=true still causes the VerifyCertifiate processes to occur

### ISSUE THREAD

**Issue by RyeNCode** [dotnet/SqlClient#2108]
### Describe the bug

When using `"Encrypt=true"` or the default of `true`, and `"TrustServerCertificate=true"` the server certificate is still verified even though the results are not used.  This caused us a problem as the Certificate Revocation List check was timing out causing a 15s delay in the initial connection of a connection pool.

The `TrustServerCertificate` flag is used as `_validateCert` within the `ValidateServerCertificate` handler passed as a callback to the `SNISslStream` constructor's `RemoteCertificateValidationCallback` parameter.
However the callback is performed after the SSL validation, and the validation results are used or ignored depending on the `TrustServerCertificate` value.

If you are seeing an exception, include the full exceptions details (message and stack trace).
_No Exception_

### To reproduce

Requirements:
* SQL Server using a certificate with a Distribution Point Name URL that does not respond within 15s.
* In my case the behaviour **did not** manifest on a Windows client
* In my case the behaviour **did** manifest on a Linux Azure WebApp
```c#
SqlConnection.ClearAllPools();
var con = new SqlConnection("server=your.server.com;Encrypt=true;TrustServerCertificate=true;");
con.Open();  //Will take 15s + actual connection overhead
con.Close();
```

### Expected behaviour
I would expect that TrustClientCertificate would actually cause the Verify Certificate process to be skipped during the SslStream opening.

### Further technical details
Microsoft.Data.SqlClient version: 5.1.1
.NET target: NetCore 6 & 7
SQL Server version: Microsoft SQL Server Enterprise (64-bit) 13.0.7016.1
Operating system: Azure Linux WebApp `uname -srvmpio
Linux 5.15.116.1-1.cm2 #1 SMP Wed Jun 21 23:17:09 UTC 2023 x86_64 unknown unknown GNU/Linux`

**Additional context**
As mentioned earlier, when testing the same code to the same SqlServers, when connecting from my developer machine, the errant validation did not seem to manifest. I suspect some difference via the Linux/Windows. I did try the `Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows` but could not detect any difference.
I could not discern any way to diagnose this via only code. Using various listeners I could not detect the SslStream's CRL check callout. I could only see it via Azure Application Insights as a failed HTTP request to the CRL identified in the server certificate prior to the SqlConnection opening call.

I believe this is what happens:
![mermaid-diagram-2023-07-27-134921](https://github.com/dotnet/SqlClient/assets/8312419/17ca6596-868a-43e5-874e-98382a35c153)

**Comment by JRahnama** [comment]
@RyeNCode I did a quick repro on Windows with `Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows` and as you mentioned issue did not happen. It all goes back to [EnableSsl](https://github.com/dotnet/SqlClient/blob/a709006541a1169f861ea2c3c6aeb5596cd1c286/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs#L611) function in SNITcpHandle, assuming it should be TCP since it is Azure, where it assigns the value for _validateCert.  I do not expect different behavior from the same code on different platform since it is the same source on Unix and Windows when the AppContext switch is applied. I will take a deeper look on how `option` is populated on EnableSsl. 

Is it possible for you to track that function to see what value is assigned to it? EventSource does not capture anything in that area. It will also be a good place to add some tracing from our side.

**Comment by RyeNCode** [comment]
Hello @JRahnama 
I've spent a few days a) learning how to do docker-compose debugging b) configuring a personal certificate that matched the conditions observed at work (it wasn't feasible to test with the actual cert) and c) writing out a test.

TLDR: The issue is rather with the "authorityInfoAccess = caIssuers;URI:" extension value being called, not the CRL (my mistake! they were the same URI in my original sample); The current SslClient does not allow us to pass options to the object that would control how the OCSP validation is done or not.

I believe the changes from [Enable server-side OCSP stapling on Linux · dotnet/runtime@6a8f16e](https://github.com/dotnet/runtime/commit/6a8f16e54fe2f91ef6457c94cab830f3c31baf9d) will allow the modification of the `SslStreamCertificateContext` of the `SslClientAuthenticationOptions` that can be passed to the `AuthenticateAsClient` method of `SslStream` and it could then be set to NOT verify this way: `SetNoOcspFetch(false)`  This looks like it's in PreRelease state currently

Either way, the testing I did was like this.
Precondition:
* SQL Server with following extensions on the Cert: 
  ```
  X509v3 CRL Distribution Points: 
     Full Name:
       URI:http://shuggoth.home.local/crll
  Authority Information Access: 
    CA Issuers - URI:http://shuggoth.home.local/info
  ```
* The VS Docker container can network out to `hastur.home.local` as the SqlServer with the encrypted connection and `shuggoth.home.local` is a http server.  I did not create a resources at `/crl` or `/info` as they are not needed. I did track which ones were access though on the host logs.

Using the following test code in the Microsoft.Data.SqlClient.DockerLinuxTest project:
```c#
using System;
using System.Collections.Generic;
using System.Diagnostics.Tracing;
using System.Linq;
namespace Microsoft.Data.SqlClient.DockerLinuxTest
{
    public class SqlClientListener : EventListener
    {
        List<string> otherSources = new();
        List<string> events = new();
        readonly List<System.Tuple<string, EventLevel, EventKeywords>> sourcesOfInterest = new()
        {

            new ( "Microsoft.Data.SqlClient.EventSource", EventLevel.Informational, (EventKeywords)(2048|512|4|2)),
            new ( "Private.InternalDiagnostics.System.Net.Http", EventLevel.Verbose, (EventKeywords)(2))
        };
        protected override void OnEventSourceCreated(EventSource eventSource)
        {
            // Only enable events from SqlClientEventSource.
            var src = sourcesOfInterest.FirstOrDefault(s => s.Item1 == eventSource.Name);
            if (src != null)
            {
                // Use EventKeyWord 2 to capture basic application flow events.
                // See the above table for all available keywords.
                EnableEvents(eventSource, src.Item2, src.Item3);
            }
            else
                otherSources.Add(eventSource.Name);
        }
        protected override void OnEventWritten(EventWrittenEventArgs eventData)
        {
            var payloadIdx = eventData.PayloadNames.IndexOf("message");
            if (payloadIdx < 0)
                payloadIdx = 0;
            events.Add($"{eventData.EventSource.Name} {{{eventData.Level}/{(uint)eventData.Keywords}}} {eventData.Payload[payloadIdx]}");

        }
        public override void Dispose()
        {
            base.Dispose();
            Console.WriteLine("Sources Not Tracked:");
            foreach (var source in otherSources)
            {
                Console.WriteLine($"\t{source}");
            }
            Console.WriteLine("Events:");
            foreach (var ev in events)
            {
                Console.WriteLine($"\t{ev}");
            }
        }
    }
    class Program
    {
        static string server = "hastur.home.local";
        static string user = "ssltest";
        static string pwd = "P@ssw0rd!123";
        static void Main(string[] args)
        {
            using (SqlClientListener listener = new SqlClientListener())
            {
                using (SqlConnection sqlConnection = new SqlConnection($"Server={server}; UID={user}; PWD={pwd};Encrypt=true;TrustServerCertificate=true"))
                {
                    sqlConnection.Open();
                    Console.WriteLine($"Connected to SQL Server v{sqlConnection.ServerVersion} from {Environment.OSVersion.VersionString}");
                    sqlConnection.Close();
                }
            }
        }
    }
}
```

The following traces were recorded:
```
-------------------------------------------------------------------
You may only use the Microsoft .NET Core Debugger (vsdbg) with
Visual Studio Code, Visual Studio or Visual Studio for Mac software
to help you develop and test your applications.
-------------------------------------------------------------------
Connected to SQL Server v16.00.1000 from Unix 5.10.16.3
Sources Not Tracked:
	Microsoft-Windows-DotNETRuntime
	System.Runtime
	System.Buffers.ArrayPoolEventSource
	Microsoft-Diagnostics-DiagnosticSource
	System.Diagnostics.Eventing.FrameworkEventSource
	System.Threading.Tasks.TplEventSource
	System.Transactions.TransactionsEventSource
	System.Net.NameResolution
	Private.InternalDiagnostics.System.Net.NameResolution
	Private.InternalDiagnostics.System.Net.Sockets
	System.Net.Sockets
	Private.InternalDiagnostics.System.Net.Security
	System.Net.Security
	System.Collections.Concurrent.ConcurrentCollectionsEventSource
	System.Net.Http
Events:
	Microsoft.Data.SqlClient.EventSource {Informational/2} <prov.DbConnectionHelper.ConnectionString_Set|API> 1, 'Server=hastur.home.local; UID=ssltest;Encrypt=true;TrustServerCertificate=true'
	Microsoft.Data.SqlClient.EventSource {Informational/4} SqlConnection.Open | API | Correlation | Object Id 1, Activity Id 2d30ac79-e204-4dbd-a602-4c23c0292d30:1
	Microsoft.Data.SqlClient.EventSource {Informational/512} SqlConnection.Open | API | Correlation | Object Id 1, Activity Id 2d30ac79-e204-4dbd-a602-4c23c0292d30:1
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.SqlConfigurableRetryLogicManager.ConnectionProvider|INFO> Requested the ConnectionProvider value.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.AppConfigManager.FetchConfigurationSection|INFO>: Unable to load custom `SqlConfigurableRetryLogicConnection`. Default value of `T` type returns.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.AppConfigManager.FetchConfigurationSection|INFO>: Unable to load custom `SqlConfigurableRetryLogicCommand`. Default value of `T` type returns.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc|SqlAuthenticationProviderManager|Ctor|Info>Neither SqlClientAuthenticationProviders nor SqlAuthenticationProviders configuration section found.
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParser.Connect | SEC | Connection Object Id 4, Authentication Mode: SqlPassword
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle..ctor | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Setting server name = hastur.home.local
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle..ctor | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Connecting to serverName hastur.home.local and port 1433
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.IPv4First | INFO | IP preference : {0}
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.Connect | INFO | Connecting to IP address 192.168.65.2 and port 1433 using InterNetwork address family. Is infinite timeout: False
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle..ctor | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390 Socket opened successfully, TCP stream ready.
	Microsoft.Data.SqlClient.EventSource {Informational/2} SNIProxy.CreateConnectionHandle | Info | Session Id 57ff9876-0687-45d9-aa91-1601f4d22390, SNI Handle Type: Microsoft.Data.SqlClient.SNI.SNITCPHandle
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.CreatePhysicalSNIHandle | Info | State Object Id 1, Session Id 57ff9876-0687-45d9-aa91-1601f4d22390, ServerName hastur.home.local, Async = True
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.GetConnectionId | Info | Session Id 57ff9876-0687-45d9-aa91-1601f4d22390
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.TdsParser.Connect|SEC> Sending prelogin handshake
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.TdsParser.SendPreLoginHandshake|INFO> ClientConnectionID 57ff9876-0687-45d9-aa91-1601f4d22390, ActivityID 2d30ac79-e204-4dbd-a602-4c23c0292d30:2
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket..ctor | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 instantiated,
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Allocate | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 allocated with _headerLength 0, _dataCapacity 94
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.AppendData | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 was appended with size 94, _dataLength 94
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.WriteToStream | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _dataLength 94 written to stream.
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.Send | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Data sent to stream synchronously
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Release | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _headerLength 0 and _dataLength 94 released.
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.WritePacket | Info | Session Id 57ff9876-0687-45d9-aa91-1601f4d22390, SendAsync Result 0
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.TdsParser.Connect|SEC> Consuming prelogin handshake
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Allocate | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 allocated with _headerLength 0, _dataCapacity 4096
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.ReadFromStream | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _dataLength 54 read from stream.
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.Receive | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Data read from stream synchronously
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.ReadSyncOverAsync | Info | State Object Id 1, Session Id 57ff9876-0687-45d9-aa91-1601f4d22390
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.ReleasePacket | Info | State Object Id 1, Session Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet DataLeft 54
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Release | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _headerLength 0 and _dataLength 54 released.
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.EnableSsl | Info | Session Id 57ff9876-0687-45d9-aa91-1601f4d22390
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SslOverTdsStream.Flush | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Flushed stream
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SslOverTdsStream.Flush | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Flushed stream
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} HttpConnectionPool http://shuggoth.home.local:80
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Available HTTP/1.1 connections: 0, Requests in the queue: 1, Pending HTTP/1.1 connections: 0, Total associated HTTP/1.1 connections: 0, Max HTTP/1.1 connection limit: 2147483647.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} No available HTTP/1.1 connections; request queued.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Creating new HTTP/1.1 connection for pool.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} HttpConnection(HttpConnectionPool http://shuggoth.home.local:80)
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} isNewConnection=True
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Dequeued waiting HTTP/1.1 request.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Sending request: Method: GET, RequestUri: 'http://shuggoth.home.local/info', Version: 1.1, Content: <null>, Headers:
{
}
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Writing 49 bytes.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Received 369 bytes.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Request is fully sent.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Received response: StatusCode: 404, ReasonPhrase: 'Not Found', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Mon, 31 Jul 2023 20:34:44 GMT
  Server: Apache/2.4.57 (Unix) OpenSSL/3.0.9
  Content-Length: 196
  Content-Type: text/html; charset=iso-8859-1
}
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} isNewConnection=False
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Put connection in pool.
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.ValidateServerCertificate | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Certificate will not be validated.
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SslOverTdsStream.FinishHandshake | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Switched from encapsulation to passthrough mode
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.EnableSsl | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, SSL enabled successfully.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.TdsParser.Connect|SEC> Prelogin handshake successful
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Allocate | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 allocated with _headerLength 0, _dataCapacity 391
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.AppendData | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 was appended with size 391, _dataLength 391
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.WriteToStream | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _dataLength 391 written to stream.
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.Send | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Data sent to stream synchronously
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Release | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _headerLength 0 and _dataLength 391 released.
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.WritePacket | Info | Session Id 57ff9876-0687-45d9-aa91-1601f4d22390, SendAsync Result 0
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Allocate | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 allocated with _headerLength 0, _dataCapacity 4096
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.ReadFromStream | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _dataLength 455 read from stream.
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.Receive | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Data read from stream synchronously
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.ReadSyncOverAsync | Info | State Object Id 1, Session Id 57ff9876-0687-45d9-aa91-1601f4d22390
	Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.ReleasePacket | Info | State Object Id 1, Session Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet DataLeft 455
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNIPacket.Release | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Packet Id 1 _headerLength 0 and _dataLength 455 released.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.TdsParser.TryRun|SEC> Received login acknowledgement token
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.AppConfigManager.FetchConfigurationSection|INFO>: Unable to load custom `AppContextSwitchOverrides`. Default value of `T` type returns.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.SqlAppContextSwitchManager.ApplyContextSwitches|INFO> Entry point.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc.SqlAppContextSwitchManager.ApplyContextSwitches|INFO> Exit point.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <sc|ActiveDirectoryAuthenticationTimeoutRetryHelper|SetState|Info>State changed from NotStarted to HasLoggedIn.
	Microsoft.Data.SqlClient.EventSource {Informational/2} <prov.DbConnectionFactory.CreatePooledConnection|RES|CPOOL> 1, Pooled database connection created.
	Microsoft.Data.SqlClient.EventSource {Informational/4} Exit Scope 1
	Microsoft.Data.SqlClient.EventSource {Informational/4} SqlConnection.Close | API | Object Id 1
	Microsoft.Data.SqlClient.EventSource {Informational/512} SqlConnection.Close | API | Correlation | Object Id 1, Activity Id 2d30ac79-e204-4dbd-a602-4c23c0292d30:2, Client Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390
	Microsoft.Data.SqlClient.EventSource {Informational/4} Exit Scope 2
	Microsoft.Data.SqlClient.EventSource {Informational/4} SqlConnection.Close | API | Object Id 1
	Microsoft.Data.SqlClient.EventSource {Informational/512} SqlConnection.Close | API | Correlation | Object Id 1, Activity Id 2d30ac79-e204-4dbd-a602-4c23c0292d30:2, Client Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390
	Microsoft.Data.SqlClient.EventSource {Informational/4} Exit Scope 3
```

I was able to drop breakpoints in `TDSParser.cs` at `lines 1051 - 1053` to observer that `ShouldValidateCertificate` was correctly set to `false` and that in `lines 1554 - 1555` that the flag `TdsEnums.SNI_SSL_VALIDATE_CERTIFICATE` was not being applied to the `uint info` variable.
The `info` variable is passed through to to the `TdsParserStateObjectManaged.EnableSSL` method after having the `TdsEnums.SNI_SSL_IGNORE_CHANNEL_BINDINGSTdsEnums.SNI_SSL_IGNORE_CHANNEL_BINDINGS` flag added.
Then it's off to `SNITCPHandle.EnableSsl` method where `_validateCert` is set to false thanks to the `options` (that was `info`) variable having the relevant flag not set.  `_tlsFirst` is false, so the SslStream's AuthenticateAsClient method is called:
```c#
_sslStream.AuthenticateAsClient(_targetServer, null, s_supportedProtocols, false);
``` 
It's at this point the SslStream's code takes over. Due to the way SsStream does the validation of the cert first then notifies the calling code via the callback, it may not be possible to alter this behaviour within the SqlClient module.

I will call your attention to the specific location on the trace that shows that after the EnableSsl call and before the Callabck, the certificate is validated:
```
Microsoft.Data.SqlClient.EventSource {Informational/2} TdsParserStateObjectManaged.EnableSsl | Info | Session Id 57ff9876-0687-45d9-aa91-1601f4d22390
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SslOverTdsStream.Flush | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Flushed stream
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SslOverTdsStream.Flush | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Flushed stream
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} HttpConnectionPool http://shuggoth.home.local:80
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Available HTTP/1.1 connections: 0, Requests in the queue: 1, Pending HTTP/1.1 connections: 0, Total associated HTTP/1.1 connections: 0, Max HTTP/1.1 connection limit: 2147483647.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} No available HTTP/1.1 connections; request queued.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Creating new HTTP/1.1 connection for pool.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} HttpConnection(HttpConnectionPool http://shuggoth.home.local:80)
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} isNewConnection=True
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Dequeued waiting HTTP/1.1 request.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Sending request: Method: GET, RequestUri: 'http://shuggoth.home.local/info', Version: 1.1, Content: <null>, Headers:
{
}
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Writing 49 bytes.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Received 369 bytes.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Request is fully sent.
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Received response: StatusCode: 404, ReasonPhrase: 'Not Found', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Date: Mon, 31 Jul 2023 20:34:44 GMT
  Server: Apache/2.4.57 (Unix) OpenSSL/3.0.9
  Content-Length: 196
  Content-Type: text/html; charset=iso-8859-1
}
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} isNewConnection=False
	Private.InternalDiagnostics.System.Net.Http {Verbose/2} Put connection in pool.
	Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.ValidateServerCertificate | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Certificate will not be validated.
```

**Comment by sabbadino** [comment]
Hi, any news on this ? we are facing the same issue (version 4.1)

**Comment by RyeNCode** [comment]
Mitigation efforts:
If it is causing a problem (like it was for me), and your org has control over the revocation list url... and or the cert... you can:
* Re-issue the cert with a functioning CRL url.
* Update the CRL to fail-fast (404 etc. rather than a long connect attempt it will return quickly)
* Get the CRL url to be a proper CRL response.
If non of those are options... I dunno, maybe try to have a fake entry in the hosts entry for that CRL host that points to a service on port 80 that just fails fast.

For our org, I was able to talk the infrastructure team though getting the CRL url working again.

**Comment by sabbadino** [comment]
Thank you for the suggestions. Consider that we are on azure Linux web app , so some of your suggestions may not apply.. but I will investigate with network team .. failing fast somehow was something I was considering.

Thanks again

**Comment by krhubbr** [comment]
I had the same problem where it would always call the URI from AIA extension of the certificate within the linux pod whereas it would work fine when it was on windows.

After a long debugging session, I figured out that it looks for the cert in the below location before it decides to call AIA URI
1. /.dotnet/corefx/cryptography/x509stores/my
2. /.dotnet/corefx/cryptography/x509stores/root 

All I had to do was place the certs in pkcs12 format with pfx extension in either my or root and then it stopped calling the AIA URI and I didn't have to update the firewall rules. 

You'll also have to make sure the owner & user id of the folder matches with currentuser in the running container. 

Luckily, I was able to replicate the entire problem by following the below guide to create cert chain on windows, 
and then install SQL Server locally, added the certificate and had the container from docker desktop call the sql server.

https://jamielinux.com/docs/openssl-certificate-authority/introduction.html

You can try adding the certs using code, this will covert crt to pfx or you can use volume mounts to map above locations to where the pfx certs reside.

```Csharp
using (var store = new X509Store(StoreName.My, StoreLocation.CurrentUser, OpenFlags.ReadWrite))
{
     Console.WriteLine($"cert: Adding root");
     store.Add(new X509Certificate2("certs/root.crt"));
     Console.WriteLine($"cert: Adding intermediate");
     store.Add(new X509Certificate2("certs/intermediate.crt"));
}
```

**Comment by arellegue** [comment]
@RyeNCode, I am just wondering if you noticed in the trace you provided:

Microsoft.Data.SqlClient.EventSource {Informational/2048} SNITCPHandle.ValidateServerCertificate | INFO | Connection Id 57ff9876-0687-45d9-aa91-1601f4d22390, Certificate will not be validated.

, if you noticed that although the ValidateServerCertificate was executed by a callback, this function did not actually validate as in the trace "Certificate will not be validated" since the ShouldValidateCertificate flag was correctly set to false.

**Comment by RyeNCode** [comment]
@arellegue regarding not-validating...
it seems the actual flow of process is that it it does the validation checks, and passes back the results to the callback.
Within that callback the option to act on the results or ignore them is allowed, and in this case the results of the validation is ignored and that trace message is produced.

The issue comes where the validation itself (before the choice to ignore it or not is available) can, under the mentioned scenario, cause a 15s delay before the connection will be allowed to continue.

**Comment by arellegue** [comment]
The underlying .Net libraries is causing the 15 second delay. Thus, the driver does not have control over it.
